### ISSUE TITLE
Different uniformity loss definition and implementation?

### ISSUE THREAD

**Issue by mprzewie** [zhangq327/U-MAE#12]
Hi, I have a question about your implementation of the uniformity loss. 

In your paper, in Eq. 11, the Uniformity loss is defined as the uniformity of feature pairs that are produced from x_1, and "an independently drawn unmasked view from X1".

However, in the codebase, the actual loss seems to be calculated as the uniformity of the matrix product of the same set of features (i.e., both product components seem to come from the same features of masked images):

https://github.com/zhangq327/U-MAE/blob/main/loss_func.py#L10

Could you help me understand how one relates to another?

**Comment by zhangq327** [comment]
Hi!
We implement the uniformity loss following the spirit of negative pairs in contratsive learning. In contrastive learning, negative pairs denote the independent samples. When implementing the objective (i.e., pushing away negative pairs), contrastive learning usually considers the pairs in the same batch as the negative pairs. Consequently, when implementing the uniformity loss in our paper, we denote the unmasked views in the same batch as the independent pairs and calculate the similarity matrix between them.

Hope it helps!

**Comment by mprzewie** [comment]
Hi, thanks for your reply and the clarifications! 
I think I got confused by three things:
* I thought that in Eq. 11, x1 (minus) refers to the same image as the masked view x1.
* I got confused by the fact that in the paper you have `f(x1).T f(x1-)` whereas in the code - `features @ features.T` (row vs column notation)
* also, the matrix formed by `features @ features.T` has entries from positive pairs on the diagonal that are included in the mean, but since we normalize the features before the product, the diagonal entries are always equal to one and effectively constant. So including the positive pairs in the sum of the matrix should not affect the loss, yes?

Thank you very much!
